src/helpers/elm/ELMInfoCache.ts

```diff
@@ -0,0 +1,93 @@
+
+export interface ELMInfoCacheType {
+  rep: Repository;
+  elmJSONs: ELM[];
+  codeService: CodeService;
+  cqls: ExtractedLibrary[];
+  rootLibIdentifier: ELMIdentifier;
+  vsMap: ValueSetMap;
+  lastAccessed?: Date;
+}
+
+// Cache entry should expire if it hasn't been accessed in 10 minutes
+const CACHE_EXPIRE_TIME = 600000;
+const ELMInfoCache = new Map<string, ELMInfoCacheType>();
+
+/**
+ * Calculates cqls, elmJSONs, rootLibIdentifier, codeService, rep, and vsMap info or retrieves from cache if available and enabled
+ * @param measure {Object} the FHIR Measure that Calculation is being run on
+ * @param measureBundle {Object} the FHIR Bundle containing the Measure and supplemental resources
+ * @param valueSets {Array} the FHIR Valueset resources referenced by the Libraries within the Measure
+ * @param useElmJsonsCaching {boolean} a flag that, when set to true, enables caching of cqls, elmJSONs,
+ * rootLibIdentifier, codeService, rep, and vsMap data for quicker use in subsequent runs
+ * @returns {Object} an object containing cqls, elmJSONs, rootLibIdentifier, codeService, rep, and vsMap
+ */
+export function retrieveELMInfo(
+  measure: fhir4.Measure,
+  measureBundle: fhir4.Bundle,
+  valueSets: fhir4.ValueSet[],
+  useElmJsonsCaching?: boolean
+): ELMInfoCacheType {
+  const { id, version } = measure;
+  const key = `${id}-${version}`;
+  if (!(ELMInfoCache.get(key) && cacheEntryIsValid(ELMInfoCache.get(key)?.lastAccessed))) {
+    const vsMap = valueSetsForCodeService(valueSets);
+
+    const { cqls, rootLibIdentifier, elmJSONs } = MeasureBundleHelpers.extractLibrariesFromBundle(measureBundle);
+
+    // add expressions for collecting for all measure observations
+    measure.group?.forEach(group => {
+      group.population
+        ?.filter(
+          population => MeasureBundleHelpers.codeableConceptToPopulationType(population.code) === PopulationType.OBSERV
+        )
+        ?.forEach(obsrvPop => {
+          const msrPop = findObsMsrPopl(group, obsrvPop);
+          if (msrPop?.criteria?.expression && obsrvPop.criteria?.expression) {
+            const mainLib = elmJSONs.find(elm => elm.library.identifier.id === rootLibIdentifier.id);
+            if (mainLib) {
+              mainLib.library.statements.def.push(
+                generateELMJSONFunction(obsrvPop.criteria.expression, msrPop.criteria.expression)
+              );
+            }
+          }
+        });
+    });
+
+    const codeService = new CodeService(vsMap);
+    const rep = new Repository(elmJSONs);
+    if (useElmJsonsCaching) {
+      ELMInfoCache.set(key, { cqls, elmJSONs, rootLibIdentifier, codeService, rep, vsMap, lastAccessed: new Date() });
+    } else {
+      return { cqls, elmJSONs, rootLibIdentifier, codeService, rep, vsMap };
+    }
+  } else {
+    ELMInfoCache.set(key, { ...(ELMInfoCache.get(key) as ELMInfoCacheType), lastAccessed: new Date() });
+  }
+  return { ...(ELMInfoCache.get(key) as ELMInfoCacheType) };
```

Comment:
is there a reason that we return a new object with a destructured object inside? Can we not just return `ELMInfoCache.get(key) as ELMInfoCacheType`? 

Also I think I see why the type casting is necessary but it took me a second. Can you add a comment just saying the cast is used since we know that the cache entry is guaranteed to be set or the function will have already returned by the time we get to this return statement?

Reply:
Yeah I wasn't entirely sure if functions would manipulate this info in ways that could mess up what the cache points to, so thought it safer to just return a copy. If not tho then will def change it to a direct return